src/AppInstallerRepositoryCore/Rest/Schema/1_0/Interface.cpp

```diff
@@ -0,0 +1,202 @@
+    constexpr std::string_view ProductCode = "ProductCode"sv;
+    constexpr std::string_view Versions = "Versions"sv;
+    constexpr std::string_view Version = "version"sv;
+    constexpr std::string_view Channel = "Channel"sv;
+
+    namespace
+    {
+        utility::string_t GetJsonKeyNameString(std::string_view nodeName)
+        {
+            return utility::conversions::to_string_t(nodeName.data());
+        }
+
+        web::json::value GetSearchBody(const SearchRequest& searchRequest)
+        {
+            // TODO: Use search request to construct search body.
+            UNREFERENCED_PARAMETER(searchRequest);
+
+            web::json::value json_body;
+            json_body[GetJsonKeyNameString(FetchAllManifests)] = web::json::value::string(L"true");
+
+            return json_body;
+        }
+
+        std::string GetRestAPIBaseUri(std::string restApiUri)
+        {
+            if (restApiUri.back() == '/')
+            {
+                restApiUri.pop_back();
+            }
+
+            return restApiUri;
+        }
+
+        utility::string_t GetSearchEndpoint(const std::string& restApiUri)
+        {
+            std::string fullSearchAPI = restApiUri;
+            return utility::conversions::to_string_t(fullSearchAPI.append(ManifestSearchPostEndpoint));
+        }
+
+        utility::string_t GetManifestByVersionEndpoint(
+            const std::string& restApiUri, const std::string& packageId, const std::string& version, const std::string& channel)
+        {
+            std::string versionEndpoint = restApiUri;
+            versionEndpoint.append(ManifestByVersionAndChannelGetEndpoint).append(packageId);
+
+            // Add Version Query param
+            if (!version.empty())
+            {
+                versionEndpoint.append("?version=").append(version);
+            }
+
+            // Add Channel Query param
+            if (!version.empty())
+            {
+                versionEndpoint.append("&channel=").append(channel);
+            }
+
+            return utility::conversions::to_string_t(versionEndpoint);
+        }
+
+        bool IsStringWhitespace(const std::string& value)
+        {
+            for (size_t i = 0; i < value.length(); i++)
+            {
+                if (!std::isspace(value[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        std::optional<std::string> GetStringFromJsonStringValue(const web::json::value& value)
+        {
+            if (value.is_null() || !value.is_string())
+            {
+                return {};
+            }
+
+            std::string result = Utility::Trim(utility::conversions::to_utf8string(value.as_string()));
+
+            if (Utility::IsEmptyOrWhitespace(result))
+            {
+                return {};
+            }
+
+            return result;
+        }
+    }
+
+    Interface::Interface(std::string restApi)

```

Comment:
>std::string [](start = 25, length = 11)

const & no need for move below

Reply:
Actually given how `GetRestAPIBaseUri` is written, taking in a value and moving it to that function was correct.  I think it is likely not to matter who makes the copy, so any of the patterns could work in practice.  But the higher you can move the forced copy up the stack, the more efficient you can be usually.

---
In reply to: [594717708](https://github.com/microsoft/winget-cli/pull/765#discussion_r594717708) [](ancestors = 594717708)